TST add global random seed to sklearn/feature_selection/tests/test_rfe.py

ClaudioSalvatoreArcidiacono opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
One step closer towards #22827.

#### What does this implement/fix? Explain your changes.
add global random seed to `sklearn/feature_selection/tests/test_rfe.py`

#### Any other comments?
For the test `test_rfe_wrapped_estimator` I had to change the number of selected features from `5` to `2` for RFECV because of the random seed change.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
add global random seed to test_rfe

Commit:
Update tests expected values

Change expected_n_features in test_rfe_wrapped_estimator
from 2 to 5 in the case of RFECV

Micky774 commented:
Hey there @ClaudioSalvatoreArcidiacono, thanks for the PR! Right now it seems you've updated _every_ random seed to use the `global_random_seed`, however this is undesirable since it leads to instability during testing if we hard-code a value such as `expected_n_features`. It's fine in cases where we know its behavior on one seed and don't need to scan multiple seeds to verify behavior. Furthermore it drastically increases the time taken to test when setting `SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all"`.

In general we want to be very selective with which tests we use the `global_random_seed` fixture for. Could you make a second pass and focus on:
1. Tests which explore mathematical properties
2. Tests which do not rely on (admittedly brittle) hard-coded values? Computed values are fine.

Please don't hesitate to ping me if you have any questions or concerns :)

Commit:
Review tests that need global_random_seed

Removed global random seed from the tests:
- test_rfecv_verbose_output
- test_rfe_cv_n_jobs
- test_rfe_cv_groups
- test_rfe_wrapped_estimator
- test_rfe_importance_getter_validation

Commit:
Merge branch 'main' into tst-use-global_random_seed-in-feature_selection-tests-test_rfe

ClaudioSalvatoreArcidiacono commented:
Hi @Micky774, thanks a lot for your comments! Now I better understood the purpose of this change.

I have reviewed the tests and I have left the hardcoded seed values for tests where I thought adding the global random seed was not adding much value. Please find the list of the tests where I have not used the `global_random_seed` fixture with explanation below.

- **test_rfecv_verbose_output:** this test checks that something is logged when the verbosity level is set to `1`. So no mathematical property is checked there.
- **test_rfe_cv_n_jobs:** the purpose of this test is to verify that `RFECV` works with `n_jobs` > 1. So, for this reason, I have left the hardcoded random seed.
- **test_rfe_cv_groups:**  the purpose of this test is to verify that `RFECV` works with `GroupKFold`. So, for this reason, I have left the hardcoded random seed.
- **test_rfe_wrapped_estimator:** the purpose of this test is to verify that `RFE` and `RFECV` support a custom importance getter, so no real mathematical property is checked here. Moreover, the check `sel.support_.sum() == expected_n_features` is unstable and the number of selected features highly changes when the seed changes.
- **test_rfe_importance_getter_validation:** Here we mainly test input validation, so the global random state is not needed.

Commit:
Run black

Micky774 committed:
Merge branch 'main' into tst-use-global_random_seed-in-feature_selection-tests-test_rfe

jeremiedbb committed:
cannot use global random seed on some tests
that rely on being useless for the learning

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/ClaudioSalvatoreArcidiacono/24142

jeremiedbb committed:
check [all random seeds]
test_rfecv_cv_results_size
test_rfe_min_step
test_number_of_subsets_of_features
test_rfe_cv_n_jobs
test_rfecv_std_and_mean

jeremiedbb reviewed (approved):
Thanks for the PR @ClaudioSalvatoreArcidiacono. LGTM.
(some tests rely on randomly generated features having no impact on the fit. This can't be guaranteed for all random seeds so I removed the used of global_random_seed for them and added a comment instead)
